### DISCUSSION TITLE
Trained AI model not working as expected on the AI deck.

### DISCUSSION THREAD

**Discussion by Kiesmith** [bitcraze/discussions#587] [Q&A]
I managed to train the AI on miniconda with the latest TensorFlow. I just tested this AI and after flashing it onto the deck and I realised that the AI I tested is not working.

I only changed the dimensions 224x224 (in the trainer.py code). It it formatted correctly in the training dataset, and it does return things in the console window and say "object detected", but it only seems to constantly say that when it looks at anything. the dataset I used to train the AI is this: https://www.kaggle.com/datasets/constantinwerner/human-detection-dataset

I understand that the code converts images to greyscale.

Any advice on this or do u know the solution?
Also: I have a pretrained model that I made myself that does actually work on my own camera. I have been trying to fit that one on instead. I have converted it to a tensorflow lite model but I havent quantized it.

Is it possible to only use a non-quantized version? I also get errors when I try to quantasize my none-related model (make it into a "_q.tflight" type).

Trained AI prediction output: Only predicts one thing and fails to predict the other. 

Thanks,
Kieran

**Comment by knmcguire** [comment]
Hi!  I think more information is needed first.

So could you go more through the process of how you tried to convert the NN for the AIdeck? Are you using the greenwaves' gapflow, or something else? Could you perhaps share some of the code for generating the NN?

**Comment by Kiesmith** [comment]
Since it is the classification, I flash the model across by the crazyradio dongle.

**This is my guide**

After implementing the greenwave technologies part so I can run the AI stuff, I train the model through the classification example: 
![image](https://user-images.githubusercontent.com/72498028/221210297-e12d1d3d-4e51-4e5b-b86f-36e6f5f322f0.png)

I moved my human detection dataset into the dataset folder and categorised it as required (the txt in that directory told me how to format the dataset).

**1)** Using miniconda I run: **python train_classifier.py**

**2)** From there, I run the rest of the command lines shown in the image to the bottom.


**Note:** Since I am flashing, instead of doing the cfloader command,  I don't use this command:
**docker run --rm -v ${PWD}:/module --device /dev/ttyUSB0 --privileged -P bitcraze/aideck tools/build/make-example [example-directory] flash**

I use flash version instead: 
**cfloader flash aideck_gap8_wifi_img_streamer_with_ap.bin deck-bcAI:gap8-fw -w [CRAZYFLIE_URI]**

Which turns out to be: 
**COMMAND: $ cfloader flash examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7** (I used this command for both the docker parts).

**Reply by knmcguire** [reply]
Thanks for the extra info! Have you changed anything to the classifier c code ? Currently the classifier used in the example has 2 outputs (background and packet), but does yours only have 1?

**Reply by Kiesmith** [reply]
It shows both packets and background. I did change it to "nothing" and "human". 

I tried pointing it at myself. it rarely says "human". 

![image](https://user-images.githubusercontent.com/72498028/222584594-32f19519-053e-4168-80d2-8898daec3faa.png)


So far just aiming the camera at myself a little bit and trying different poses that aren't really poses seems to detect me but the majority of it is saying I'm not there. 

**Overview**

**1:** When pointed at towards human, rarely detects and says "human" on the console tab. However: Also says "Human" to anything else randomly. I pointed it to my chair and it detected the chair as a human. Detected the floor as a human. Detected a closeup of the leather side of my chair as a "Human". 

**2:** Does not seem to detect anything when pointed away from human, that is: covering camera so it cannot see anything still says "nothing".

**Reply by hmllr** [reply]
It is really hard to say if the problem is the different camera, somewhere in the deployment process or even the quantization.
To debug this, I would run it on images where you know the output - so not using the camera but directly images you also tested on in python. This way you can exclude that it is the camera differences.
You find an example on how to read images from files on gap8 see e.g. https://github.com/GreenWaves-Technologies/gap_sdk/tree/master/examples/gap8/nn/nntool/mnist (not sure if those examples are up to date, but it will hopefully at least give you some inspiration on how to start).
Hope this helps,
Hanna
